### DISCUSSION TITLE
AI based auto generated tags/categorisation

### DISCUSSION THREAD

**Discussion by sovetski** [navidrome/navidrome#3145] [Ideas]
Hello,

I think I am not the only one who enjoys listening to music from various countries and genres. Sometimes, I just want to listen to music from a specific country or a particular genre.

No platform currently offers this feature, not Spotify, not Apple Music, etc. What I do is create a playlist for each country, and then sub-playlists for each category, e.g., French => Rap, Calm, English => Rap, Calm, etc., etc.

This is unmanageable because I keep repeating myself, and the main problem is that I can't say, "I want to listen to rap playlists in all languages." I have to choose the specific playlist each time.

Is it possible to add some sort of AI or another feature that detects this kind of information from each track and adds tags according to the language(s), genre, etc., so that we can choose to listen to music that only contains the tags we selected?

**Comment by deluan** [comment]
Although I think this is a great idea, this would require access to a local LLM (heavy) or an API like OpenAI's (paid). 

Once we introduce plugins to Navidrome, I may spend some time writing one that uses OpenAI API to do this, but it will require the user to pay for the API usage.

**Comment by kabachuha** [comment]
More lightweight CLIP/Sound Encoders/LPIPS-like, but for sounds, semantic similarity retrieval/recommendations would be nice to have too and are easier to implement

Looking forward for the plugins, as much more interesting stuff can be ahead for those who have more potent home server stations

**Reply by kabachuha** [reply]
There are even classic algorithms for music similarity findings using tonality/modulations detection, for example in musicpy

https://github.com/Rainbow-Dreamer/musicpy/wiki/The-algorithm-to-analyze-tonality-and-modulations-in-a-piece-of-music

https://github.com/Rainbow-Dreamer/musicpy/blob/master/musicpy/algorithms.py#L125

**Reply by deluan** [reply]
Plugins will probably be implemented with WASM, so we can potentially create a plugin using bliss-rs (written in Rust): #2259
